### PR TITLE
[Tables] Fix for async retry policy

### PR DESCRIPTION
**Pull request by annatisch**

### CODE REVIEW

File: sdk/tables/azure-data-tables/tests/test_retry.py [hunk 1]
```diff
@@ -0,0 +1,188 @@
...
+    def test_retry_on_timeout(self, tables_storage_account_name, tables_primary_storage_account_key):
+        retry = ExponentialRetry(initial_backoff=1, increment_base=2)
+        self._set_up(tables_storage_account_name, tables_primary_storage_account_key, retry_policy=retry, default_table=False)
+
+        new_table_name = self.get_resource_name('uttable')
+        callback = ResponseCallback(status=201, new_status=408).override_status
+
+        try:
+            # The initial create will return 201, but we overwrite it and retry.
+            # The retry will then get a 409 and return false.
```
Comment (seankane-msft) [inline sdk/tables/azure-data-tables/tests/test_retry.py:112]: the callback says `new_status=408` but your comment says the retry is a `409`, just for clarification is it expecing a 408 or 409?

File: sdk/tables/azure-data-tables/tests/test_retry.py [hunk 2]
```diff
@@ -0,0 +1,188 @@
...
+            with self.assertRaises(ResourceExistsError):
```
Comment (seankane-msft) [inline sdk/tables/azure-data-tables/tests/test_retry.py:113]:
```suggestion
            with pytest.raises(ResourceExistsError):
```

File: sdk/tables/azure-data-tables/tests/test_retry.py [hunk 3]
```diff
@@ -0,0 +1,188 @@
...
+    def test_retry_on_server_error(self, tables_storage_account_name, tables_primary_storage_account_key):
+        self._set_up(tables_storage_account_name, tables_primary_storage_account_key, default_table=False)
+        try:
+            # Force the create call to 'timeout' with a 408
+            callback = ResponseCallback(status=201, new_status=500).override_status
+
+            new_table_name = self.get_resource_name('uttable')
+            # The initial create will return 201, but we overwrite it and retry.
+            # The retry will then get a 409 and return false.
+            with self.assertRaises(ResourceExistsError):
```
Comment (seankane-msft) [inline sdk/tables/azure-data-tables/tests/test_retry.py:95]:
```suggestion
            with pytest.raises(ResourceExistsError):
```

File: sdk/tables/azure-data-tables/tests/test_retry.py [hunk 4]
```diff
@@ -0,0 +1,188 @@
...
+        new_table_name = self.get_resource_name('uttable')
+        callback = ResponseCallback(status=201, new_status=408).override_status
+
+        def assert_exception_is_present_on_retry_context(**kwargs):
+            self.assertIsNotNone(kwargs.get('response'))
+            self.assertEqual(kwargs['response'].status_code, 408)
+        try:
+            # The initial create will return 201, but we overwrite it and retry.
+            # The retry will then get a 408 and return false.
+            with self.assertRaises(ResourceExistsError):
```
Comment (seankane-msft) [inline sdk/tables/azure-data-tables/tests/test_retry.py:134]:
```suggestion
            with pytest.raises(ResourceExistsError):
```

File: sdk/tables/azure-data-tables/tests/test_retry.py [hunk 5]
```diff
@@ -0,0 +1,188 @@
...
+        self._set_up(
+            tables_storage_account_name,
+            tables_primary_storage_account_key,
+            retry_policy=retry,
+            transport=retry_transport,
+            default_table=False)
+    
+        new_table_name = self.get_resource_name('uttable')
+        try:
+            with self.assertRaises(AzureError) as error:
```
Comment (seankane-msft) [inline sdk/tables/azure-data-tables/tests/test_retry.py:154]:
```suggestion
            with pytest.raises(AzureError) as error:
```

File: sdk/tables/azure-data-tables/tests/test_retry_async.py [hunk 6]
```diff
@@ -0,0 +1,188 @@
...
+    async def test_retry_on_server_error_async(self, tables_storage_account_name, tables_primary_storage_account_key):
+        await self._set_up(tables_storage_account_name, tables_primary_storage_account_key, default_table=False)
+        try:
+            # Force the create call to 'timeout' with a 408
+            callback = ResponseCallback(status=201, new_status=500).override_status
+
+            new_table_name = self.get_resource_name('uttable')
+            # The initial create will return 201, but we overwrite it and retry.
+            # The retry will then get a 409 and return false.
+            with self.assertRaises(ResourceExistsError):
```
Comment (seankane-msft) [inline sdk/tables/azure-data-tables/tests/test_retry_async.py:93]:
```suggestion
            with pytest.raises(ResourceExistsError):
```

File: sdk/tables/azure-data-tables/tests/test_retry_async.py [hunk 7]
```diff
@@ -0,0 +1,188 @@
...
+        retry = ExponentialRetry(initial_backoff=1, increment_base=2)
+        await self._set_up(tables_storage_account_name, tables_primary_storage_account_key, retry_policy=retry, default_table=False)
+
+        new_table_name = self.get_resource_name('uttable')
+        callback = ResponseCallback(status=201, new_status=408).override_status
+
+        try:
+            # The initial create will return 201, but we overwrite it and retry.
+            # The retry will then get a 409 and return false.
+            with self.assertRaises(ResourceExistsError):
```
Comment (seankane-msft) [inline sdk/tables/azure-data-tables/tests/test_retry_async.py:111]:
```suggestion
            with pytest.raises(ResourceExistsError):
```

File: sdk/tables/azure-data-tables/tests/test_retry_async.py [hunk 8]
```diff
@@ -0,0 +1,188 @@
...
+        new_table_name = self.get_resource_name('uttable')
+        callback = ResponseCallback(status=201, new_status=408).override_status
+
+        def assert_exception_is_present_on_retry_context(**kwargs):
+            self.assertIsNotNone(kwargs.get('response'))
+            self.assertEqual(kwargs['response'].status_code, 408)
+        try:
+            # The initial create will return 201, but we overwrite it and retry.
+            # The retry will then get a 408 and return false.
+            with self.assertRaises(ResourceExistsError):
```
Comment (seankane-msft) [inline sdk/tables/azure-data-tables/tests/test_retry_async.py:132]:
```suggestion
            with pytest.raises(ResourceExistsError):
```

File: sdk/tables/azure-data-tables/tests/test_retry_async.py [hunk 9]
```diff
@@ -0,0 +1,188 @@
...
+        await self._set_up(
+            tables_storage_account_name,
+            tables_primary_storage_account_key,
+            retry_policy=retry,
+            transport=retry_transport,
+            default_table=False)
+    
+        new_table_name = self.get_resource_name('uttable')
+        try:
+            with self.assertRaises(AzureError) as error:
```
Comment (seankane-msft) [inline sdk/tables/azure-data-tables/tests/test_retry_async.py:152]:
```suggestion
            with pytest.raises(AzureError):
```

**Review by seankane-msft — APPROVED** [review]
Some nits, otherwise LGTM

File: sdk/tables/azure-data-tables/tests/test_retry.py (hunk 1 above)
Comment (annatisch) [reply sdk/tables/azure-data-tables/tests/test_retry.py:112]: Actually both - I will update the comments to be clearer.

**Pull request merged**